### PR TITLE
Dashboard: Externalize WordPress dependencies

### PR DESCRIPTION
**Pull request by sirreal**
Uses webpack plugin to externalize WordPress script dependencies, produce a manifest, and read from the manifest to register the scripts.

Use `lodash` "member" imports, update eslint rule accordingly. The webpack plugin currently detects `lodash` imports, but not `lodash/method` imports.

The produced deps JSON is:

```json
["lodash","moment","react","react-dom"]
```

```
                Asset      Size  Chunks                    Chunk Names
      admin.deps.json  39 bytes       0  [emitted]         admin
 admin.dops-style.css  76.2 KiB       0  [emitted]         admin
             admin.js  1.88 MiB       0  [emitted]  [big]  admin
     static.deps.json  27 bytes       1  [emitted]         static
static.dops-style.css  11.2 KiB       1  [emitted]         static
            static.js   1.2 MiB       1  [emitted]  [big]  static
Entrypoint admin [big] = admin.dops-style.css admin.js admin.deps.json
Entrypoint static [big] = static.dops-style.css static.js static.deps.json
chunk    {0} admin.dops-style.css, admin.js, admin.deps.json (admin) 3.1 MiB [entry] [rendered]
chunk    {1} static.dops-style.css, static.js, static.deps.json (static) 1.61 MiB [entry] [rendered]
```

Compare admin.js:

| `master` | this branch | Δ
| --- | --- | --- |
| 2.25 MB | 1.88 MB | -370 KB / -16% |

Supersedes #11655 

#### Testing instructions:
* The dashboard should continue work as expected with no regressions.

#### Proposed changelog entry for your changes:
* Improve Jetpack dashboard performance by leveraging shared WordPress scripts.

### CODE REVIEW

**Comment by ockham** [comment]
`eslint` still complaining about a few lodash imports:

```
/home/travis/build/Automattic/jetpack/webpack.config.extensions.js
  9:7  error  Import members from the full Lodash module  lodash/import-scope
```

```
/home/travis/build/Automattic/jetpack/extensions/blocks/publicize/form.js
  11:1  error  Import members from the full Lodash module  lodash/import-scope
```

File: tools/builder/react.js [hunk 1]
```diff
@@ -171,6 +171,10 @@ function buildStatic( done ) {
 	global.window = window;
 	global.document = document;
 	global.navigator = window.navigator;
+	global.React = require( 'react' );
+	global.ReactDOM = require( 'react-dom' );
+	global.lodash = require( 'lodash' );
+	global.moment = require( 'moment' );
```
Comment (ockham) [inline tools/builder/react.js]: Uh is this how we externalize these dependencies for the React dashboard :question:

Comment (sirreal) [reply tools/builder/react.js]: I believe we're using React to build some static HTML in a way I don't fully understand. Because these dependencies are now external, we need to make sure they're globals like they would be in the browser.

If you stop `yarn build-client` before it cleans up these files, you'll see `static.deps.json`:

```json
["lodash","moment","react"]
```

Maybe @zinigor can look this bit over 🙇

**Comment by sirreal** [comment]
Lodash imports fixed

**Comment by simison** [comment]
Will this essentially mean that we're making React dependency depend on "any" version, or will Jetpack's dependency to specific WP core versions lock it down enough for us?

When React is used via wp.element and friends, feels like there's stronger contract for APIs being deprecated between WordPress versions.

At any rate, if this becomes and issue, we can always start including React again in the bundle?

**Comment by sirreal** [comment]
Good questions, @simison 👍 

> Will this essentially mean that we're making React dependency depend on "any" version, or will Jetpack's dependency to specific WP core versions lock it down enough for us?

The dashboard in WP Admin will use whatever version core is providing. That seems good.

The static HTML page will use whatever React we depend on, that seems OK, but it would be nice if it used the same version as core.

All that said, recent versions of React are very stable and I don't expect any breaking changes to hit us without a lot of notice.

What we will need to ensure going forward is that when we adopt new React features like hooks, the core versions we're supporting must be on the right version of React.

> When React is used via wp.element and friends, feels like there's stronger contract for APIs being deprecated between WordPress versions.

We could move to using `wp-element` for the dashboard. It should be a question of adding the babel transforms and removing a bunch of React imports. The version question is still there but the `@wordpress` org can figure it out and provide information on how to go about adopting.

> At any rate, if this becomes and issue, we can always start including React again in the bundle?

Definitely.

**Review by ockham — DISMISSED** [review]
This is looking great -- thanks a lot for working on it, Jon!

Seems to work fine in my testing, too!

**Review by ockham — DISMISSED** [review]
Still good :+1:

File: _inc/client/at-a-glance/activity.jsx [hunk 2]
```diff
@@ -6,8 +6,7 @@ import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
 import DashItem from 'components/dash-item';
 import { translate as __ } from 'i18n-calypso';
-// import get from 'lodash/get';
-// import includes from 'lodash/includes';
+// import { get, includes } from 'lodash';
```
Comment (oskosk) [inline _inc/client/at-a-glance/activity.jsx]: why keep this?

Comment (sirreal) [reply _inc/client/at-a-glance/activity.jsx]: Seems out of scope, presumably was kept for a reason. I'm fine either way.

Comment (sirreal) [reply _inc/client/at-a-glance/activity.jsx]: From #8311 - seems like a pending todo that's waiting on rewind.

**Comment by sirreal** [comment]
Rebased to fix conflicts.

File: _inc/client/at-a-glance/activity.jsx (hunk 2 above)
Comment (sirreal) [reply _inc/client/at-a-glance/activity.jsx]: Oh, be sure to look at the rest of the file. This is part of a bunch of code that was commented out 🙂

File: tools/builder/react.js (hunk 1 above)
Comment (zinigor) [reply tools/builder/react.js]: Yes, we are using React in the build phase to create several HTML files that are used as loading placeholders. This allows us to use the same components (with limitations) as we do for the rest of the admin page. The method that prepares DOM and requires the JS file is here: https://github.com/Automattic/jetpack/blob/master/tools/builder/react.js#L166

I'm not sure how we're using WordPress dependencies here, but yes, we need to have all of these for the HTML builder entry point to be able to use them.

File: _inc/lib/admin-pages/class.jetpack-react-page.php [hunk 3]
```diff
@@ -136,10 +136,16 @@ function page_admin_scripts() {
...
+		$script_dependencies = file_exists( $script_deps_path )
+			? json_decode( file_get_contents( $script_deps_path ) )
+			: array();
+		$script_dependencies[] = 'wp-polyfill';
+
 		wp_enqueue_script(
 			'react-plugin',
 			plugins_url( '_inc/build/admin.js', JETPACK__PLUGIN_FILE ),
-			array( 'wp-i18n', 'wp-polyfill' ),
+			$script_dependencies,
```
Comment (simison) [inline _inc/lib/admin-pages/class.jetpack-react-page.php]: Do you know  why `wp-i18n` was a dependency here?

Looks  like admin panel  is using `import { translate as __ } from 'i18n-calypso';`  instead. 🤔

Comment (simison) [reply _inc/lib/admin-pages/class.jetpack-react-page.php]: Looks like it came from https://github.com/Automattic/jetpack/pull/11225

The new `admin.deps.json` doesn't include it so seems like it wasn't needed:
```json
["lodash","moment","react","react-dom"]
```

**Review by simison — DISMISSED** [review]

File: _inc/lib/admin-pages/class.jetpack-react-page.php (hunk 3 above)
Comment (sirreal) [reply _inc/lib/admin-pages/class.jetpack-react-page.php]: > Do you know why `wp-i18n` was a dependency here?
> 
> Looks like admin panel is using `import { translate as __ } from 'i18n-calypso';` instead. 🤔

TL;DR - pretty sure it's safe to remove that dependency.

---

I suspect it's related to this:

https://github.com/Automattic/jetpack/blob/17221d7fbd37799508f5b1f43c41e46b7dea2a0b/_inc/lib/admin-pages/class.jetpack-react-page.php#L147

That ends up hitting [this](https://github.com/WordPress/WordPress/blob/49351e0578797a70c592802674ff2b120c367f21/wp-includes/class.wp-scripts.php#L498-L522) which adds the dependency to the script:

```php
	public function set_translations( $handle, $domain = 'default', $path = null ) {
		if ( ! isset( $this->registered[ $handle ] ) ) {
			return false;
		}

		/** @var \_WP_Dependency $obj */
		$obj = $this->registered[ $handle ];

		if ( ! in_array( 'wp-i18n', $obj->deps, true ) ) {
			$obj->deps[] = 'wp-i18n';
		}

		return $obj->set_translations( $domain, $path );
	}
```

which is necessary, because [printing the locales depends](https://github.com/WordPress/WordPress/blob/60670888486617c6eeba7f6d038dd821941bbda3/wp-includes/class.wp-scripts.php#L549-L555) on `wp.i18n`:

```
		$output = <<<JS
( function( domain, translations ) {
	var localeData = translations.locale_data[ domain ] || translations.locale_data.messages;
	localeData[""].domain = domain;
	wp.i18n.setLocaleData( localeData, domain );
} )( "{$domain}", {$json_translations} );
JS;
```

Comment (zinigor) [reply _inc/lib/admin-pages/class.jetpack-react-page.php]: You both were right about this one - this was intended to be the Settings area using wp.i18n, but it didn't happen yet. So I have pushed a new commit that would remove `set_script_translations` entirely, so that we wouldn't need `wp-i18n` as a dependency.

**Review by zinigor — APPROVED** [review]
Works well for me, thanks for the update, here's to shipping less code!

**Review by kraftbj — APPROVED** [review]
It built fine and trying out the dash, didn't run into any problems or errors. Changed the site's language and confirmed translations still loaded.
